semgrep-core/src/engine/Match_extract_mode.ml

```diff
@@ -0,0 +1,374 @@
+      Common.map
+        (fun (m : Pattern_match.t) ->
+          { m with file; range_loc = Common2.pair map_loc m.range_loc })
+        mr.matches;
+    errors =
+      Common.map
+        (fun (e : Semgrep_error_code.error) -> { e with loc = map_loc e.loc })
+        mr.errors;
+    skipped_targets =
+      Common.map
+        (fun (st : Output_from_core_t.skipped_target) ->
+          { st with path = (if st.path = tmpfile then file else st.path) })
+        mr.skipped_targets;
+    profiling = { mr.profiling with Report.file };
+  }
+
+(*****************************************************************************)
+(* Main logic *)
+(*****************************************************************************)
+
+let extract_and_concat erule_table xtarget rule_ids matches =
+  matches
+  (* Group the matches within this file by rule id *)
+  |> collect (fun m m' -> m.Pattern_match.rule_id = m'.Pattern_match.rule_id)
+  |> Common.map (fun matches -> nonempty_to_list matches)
+  (* Convert matches to the extract metavariable / bound value *)
+  |> Common.map
+       (List.filter_map (fun m ->
+            match extract_of_match erule_table m with
+            | Some ({ mode = `Extract { Rule.extract; _ }; id = id, _; _ }, None)
+              ->
+                report_unbound_mvar id extract m;
+                None
+            | Some (r, Some mval) -> Some (r, mval)
+            | None -> None))
+  (* Factor out rule *)
+  |> List.filter_map (function
+       | [] -> None
+       | (r, _) :: _ as xs -> Some (r, Common.map snd xs))
+  (* Convert mval match to offset of location in file *)
+  |> Common.map (fun (r, mvals) ->
+         ( r,
+           List.filter_map
+             (fun mval ->
+               let offsets = offsets_of_mval mval in
+               if Option.is_none offsets then report_no_source_range r;
+               offsets)
+             mvals ))
+  (* For every rule ... *)
+  |> Common.map (fun (r, offsets) ->
+         (* Sort matches by start position for merging *)
+         List.fast_sort (fun x y -> Int.compare x.start_pos y.start_pos) offsets
+         |> List.fold_left
+              (fun acc curr ->
+                match acc with
+                | [] -> [ curr ]
+                | last :: acc ->
+                    (* Keep both when disjoint *)
+                    if last.end_pos < curr.start_pos then curr :: last :: acc
+                      (* Filter out already contained range; start of
+                           last is before start of curr from sorting *)
+                    else if curr.end_pos <= last.end_pos then last :: acc
+                      (* Merge overlapping ranges *)
+                    else { last with end_pos = curr.end_pos } :: acc)
+              []
+         |> List.rev
+         (* Read the extracted text from the source file *)
+         |> Common.map (fun { start_pos; start_line; start_col; end_pos } ->
+                let source_file = open_in_bin xtarget.Xtarget.file in
```

Comment:
you forgot to close_in here, leading to "too many opened files errors".
(I've fixed it today).

Reply:
Whoops, sorry!